lib/Pico.php

```diff
@@ -0,0 +1,945 @@
+     */
+    public function getFileMeta()
+    {
+        return $this->meta;
+    }
+
+    /**
+     * Applies some static preparations to the raw contents of a page,
+     * e.g. removing the meta header and replacing %base_url%
+     *
+     * @param  string $rawContent raw contents of a page
+     * @return string             contents prepared for parsing
+     */
+    public function prepareFileContent($rawContent)
+    {
+        // remove meta header
+        $metaHeaderPattern = "/^(\/(\*)|---)[[:blank:]]*(?:\r)?\n"
+            . "(.*?)(?:\r)?\n(?(2)\*\/|---)[[:blank:]]*(?:(?:\r)?\n|$)/s";
+        $content = preg_replace($metaHeaderPattern, '', $rawContent, 1);
+
+        // replace %site_title%
+        $content = str_replace('%site_title%', $this->getConfig('site_title'), $content);
+
+        // replace %base_url%
+        if ($this->isUrlRewritingEnabled()) {
+            // always use `%base_url%?sub/page` syntax for internal links
+            // we'll replace the links accordingly, depending on enabled rewriting
+            $content = str_replace('%base_url%?', $this->getBaseUrl(), $content);
+        } else {
+            // actually not necessary, but makes the URLs look a little nicer
+            $content = str_replace('%base_url%?', $this->getBaseUrl() . '?', $content);
+        }
+        $content = str_replace('%base_url%', rtrim($this->getBaseUrl(), '/'), $content);
+
+        // replace %theme_url%
+        $themeUrl = $this->getBaseUrl() . basename(THEMES_DIR) . '/' . $this->getConfig('theme');
+        $content = str_replace('%theme_url%', $themeUrl, $content);
+
+        // replace %meta.*%
+        $metaKeys = array_map(function ($metaKey) {
+            return '%meta.' . $metaKey . '%';
+        }, array_keys($this->meta));
+        $metaValues = array_values($this->meta);
+        $content = str_replace($metaKeys, $metaValues, $content);
+
+        return $content;
+    }
+
+    /**
+     * Parses the contents of a page using ParsedownExtra
+     *
+     * @param  string $content raw contents of a page (Markdown)
+     * @return string          parsed contents (HTML)
+     */
+    public function parseFileContent($content)
+    {
+        $parsedown = new ParsedownExtra();
+        return $parsedown->text($content);
+    }
+
+    /**
+     * Returns the cached contents of the requested page
+     *
+     * @see    Pico::parseFileContent()
+     * @return string parsed contents
+     */
+    public function getFileContent()
+    {
+        return $this->content;
+    }
+
+    /**
+     * Reads the data of all pages known to Pico
+     *
+     * @return void
+     */
+    protected function readPages()
+    {
+        $pages = array();
+        $files = $this->getFiles($this->getConfig('content_dir'), $this->getConfig('content_ext'));
+        foreach ($files as $i => $file) {
+            // skip 404 page
+            if (basename($file) == '404' . $this->getConfig('content_ext')) {
```

Comment:
See the commented line and 77f9390 :smile:

Pico 0.9 excluded all `404.md` files in any directory from the pages list, even they weren't functional. This never was documented and is unexpected behavior - you either expect that only the global `404.md` file is removed or that Pico supports per-directory `404.md` files. The first solution would break BC, so I decided that the additional 10 LoC are it worth to resolve this problem.